### ISSUE TITLE
Reduce calls to jarDefinesClass

### ISSUE THREAD

**Issue by ijuma** [sbt/sbt#67]
jarDefinesClass involves opening a JAR and reading its entries and it's the biggest hotspot according to YourKit when profiling a test:compile invocation on a medium-sized multi-module project with many dependencies. Between 70% and 80% of the time is spent there. 

For this project, 1358(!) jars are opened in a single invocation of test:compile. Part of the problem is that the same jar is opened multiple times. For example:

opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar
opening jar: <project_root>/lib_managed/jars/commons-io/commons-io/commons-io-2.0.1.jar

Can we not do better?

By the way, externalBinaryModified is the source of too many calls to Locate.entry which eventually calls jarDefinesClass.

**Comment by harrah** [comment]
Yeah, Scott mentioned this to me at ScalaDays.

**Comment by ijuma** [comment]
Do you think it would be complicated to improve it?

**Comment by harrah** [comment]
Well, if those 26 lines correspond to 26 subprojects, I'd assume a shared index would help.  If so, it might not be too bad.

**Comment by ijuma** [comment]
13 projects \* 2 (for compile and test scope I guess) = 26.

A shared index is indeed what seemed like an obvious improvement. Great that there is no reason why that would be difficult to do.

**Comment by ijuma** [comment]
Thanks! The profile is spread around more methods now.

**Comment by harrah** [comment]
What about actual runtime?

**Comment by ijuma** [comment]
Along with the change for #66 that you suggested, SBT now consistently reports 1s for test:compile (it would be nice to have better granularity, but I removed the instrumentation I had added). Previously things would vary a lot and the best times reported were 2s and they could go to 5-6s easily. Also the CPU usage seems to be much lower now, which provides for a better experience (people tend to have other things running that need CPU, including their IDE).

I believe the reason why there was a lot of variation earlier was due to garbage collection. A lot of garbage was being generated and full GC of a 2 GB heap takes a decent amount of time. I think it would probably be a good idea to try to reduce the number of allocations in the common path.

So, are we close to 0.10.1? :)

**Comment by ijuma** [comment]
I should also mention that I've been testing on a machine with 8 cores (two quad cores). So, I believe that is also part of the reason why the differences aren't even larger (as the jars would be in the disk cache and then they would be scanned in parallel due to parallelExecution).
